Pasfmt.Cursors.pas

```diff
@@ -0,0 +1,339 @@
+  Result.FThreadCondition := SourceBreakpoint.ThreadCondition;
+end;
+
+//______________________________________________________________________________________________________________________
+
+procedure TBreakpointData.CopyTo(SourceBreakpoint: IOTABreakpoint);
+begin
+  SourceBreakpoint.DoHandleExceptions := Self.FDoHandleExceptions;
+  SourceBreakpoint.StackFramesToLog := Self.FStackFramesToLog;
+  SourceBreakpoint.DoIgnoreExceptions := Self.FDoIgnoreExceptions;
+  SourceBreakpoint.DoBreak := Self.FDoBreak;
+  SourceBreakpoint.DisableGroup := Self.FDisableGroup;
+  SourceBreakpoint.EvalExpression := Self.FEvalExpression;
+  SourceBreakpoint.LogResult := Self.FLogResult;
+  SourceBreakpoint.EnableGroup := Self.FEnableGroup;
+  SourceBreakpoint.LogMessage := Self.FLogMessage;
+  SourceBreakpoint.GroupName := Self.FGroupName;
+  SourceBreakpoint.FileName := Self.FFileName;
+  SourceBreakpoint.Enabled := Self.FEnabled;
+  SourceBreakpoint.LineNumber := Self.FLineNumber;
+  SourceBreakpoint.Expression := Self.FExpression;
+  SourceBreakpoint.PassCount := Self.FPassCount;
+  SourceBreakpoint.ThreadCondition := Self.FThreadCondition;
+end;
+
+//______________________________________________________________________________________________________________________
+
+destructor TCursors.Destroy;
+begin
+  FreeAndNil(FBookmarkOffsets);
+  FreeAndNil(FBookmarkIds);
+  FreeAndNil(FBreakpointOffsets);
+  FreeAndNil(FBreakpointData);
+  inherited;
+end;
+
+//______________________________________________________________________________________________________________________
+
+constructor TCursors.Create(Buffer: IOTAEditBuffer);
+var
+  I: Integer;
+  EditView: IOTAEditView;
+  EditPos: TOTAEditPos;
+  CharPos: TOTACharPos;
+  Bookmark: Integer;
+  DebuggerServices: IOTADebuggerServices;
+  Breakpoint: IOTASourceBreakpoint;
+begin
+  SetLength(FOffsets, Buffer.EditViewCount);
+  SetLength(FRowsFromTop, Buffer.EditViewCount);
+
+  FBookmarkOffsets := TList<Integer>.Create;
+  FBookmarkIds := TList<Integer>.Create;
+  FBreakpointOffsets := TList<Integer>.Create;
+  FBreakpointData := TList<TBreakpointData>.Create;
+
+  for I := 0 to Buffer.EditViewCount - 1 do begin
+    EditView := Buffer.EditViews[I];
+
+    EditPos := EditView.CursorPos;
+    EditView.ConvertPos(True, EditPos, CharPos);
+    FOffsets[I] := EditView.CharPosToPos(CharPos);
+    FRowsFromTop[I] := EditView.Position.Row - EditView.TopRow;
+  end;
+
+  if Buffer.EditViewCount <= 0 then begin
+    Log.Debug('Buffer has no edit views');
+    Exit;
+  end;
+
+  // bookmarks are shared between all edit views of a buffer
+  EditView := Buffer.EditViews[0];
+
+  for Bookmark := 0 to 19 do begin
+    CharPos := EditView.BookmarkPos[Bookmark];
+    if CharPos.Line <> 0 then begin
+      FBookmarkOffsets.Add(EditView.CharPosToPos(CharPos));
+      FBookmarkIds.Add(Bookmark);
+    end;
+  end;
+
+  if BorlandIDEServices.QueryInterface(IOTADebuggerServices, DebuggerServices) = S_OK then begin
```

Comment:
Same question about `BorlandIDEServices` ever not implementing `IOTADebuggerServices`. Also, `System.SysUtils.Supports` wraps `QueryInterface` and is shorter + returns a boolean out of the box - might be better to use that instead.

Reply:
Look, I don't know what the deal is with the interfaces and when they are implemented. Without any kind of compile-time assurance, where the functionality isn't essential to the plugin, I'd prefer to guard the behaviour with an explicit check like this.

I will which to `Supports`, though.

Reply:
OK, sure. Doesn't really bother me.